---
title: "Computational Workflows for Single Cell and Spatial Transcriptomics"
author: "Robert R Butler III, Ph.D."
output:
  html_document:
    toc: true
    toc_float: true
    df_print: paged
    fig_width: 10
    fig_align: center
---

```{r, load_up, include = FALSE}
# knitr options

# packages to load
library(Seurat) # uses DESeq2 and MAST
library(ggplot2)
library(data.table)
library(Giotto)
library(patchwork)

```
# Introductory concepts (lecture)


# Single-cell analysis (interactive)
To begin, this course assumes you have a basic familiarity with single-cell analysis using the [Seurat](https://satijalab.org/seurat/) package. We won't cover in huge detail the major steps in a single-cell analysis, but leave it for you to explore. Our key steps:

1. Clean the data (not included) - see tools like SoupX, DoubletFinder, etc
2. (Optional) Integrate datasets, if multiple
3. Normalize & Scale
4. Calculate Principal Components
5. Find Neighbors
6. Find Clusters
7. Run UMAP

## Prep the (cleaned) data
This will define standard settings you intend to run for the analysis
```{r, get_sc_data}
# sc <- readRDS()
sc <- pbmc_small

# do you have multiple conditions to integrate?
integ <- TRUE
# what column name contaions the groups you are integrating?
g_col <- 'groups' 
# how many principal components do you wish to use?
pcs <- 5
# what clustering resolution? 
resi <- 0.8

```

## Clean and Analyze
The bulk of the processing can be done in rapid succession, with linking `%>%` chains.
```{r, sc_process, message = FALSE, warning = FALSE}
if (integ) {
  # split by group column
  sc[["RNA"]] <- split(sc[["RNA"]], f = sc@meta.data[[g_col]])

  # Prep & PCA on each group
  sc %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData(vars.to.regress = c("letter.idents")) %>%
    RunPCA(npcs = pcs) %>%
    RunUMAP(dims = 1:pcs, reduction.name = "umap") %>%

    # feature selection and integration prep
    IntegrateLayers(
      method = HarmonyIntegration,
      new.reduction = "harmony",
      verbose = TRUE
    ) %>%
    RunUMAP(
      dims = 1:pcs, 
      reduction = "harmony", 
      reduction.name = "umap.harmony"
    ) %>%
    FindNeighbors(reduction = "harmony", dims = 1:pcs) %>%
    FindClusters(resolution = resi) %>%

    # combine again
    JoinLayers() -> sc

} else { # otherwise, ignore integration
  sc %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData(vars.to.regress = c("letter.idents")) %>%
    RunPCA(npcs = pcs) %>%
    FindNeighbors(dims = 1:pcs) %>%
    FindClusters() %>%
    RunUMAP(dims = 1:pcs) -> sc
} 

```

How did that integration turn out? We can visualize with the two UMAPs if we did integration, or just plotting the standard UMAP if we did not.
```{r, integ_plot}
if (integ) {
  p1 <- DimPlot(sc, reduction = "umap", group.by = g_col)
  p2 <- DimPlot(sc, reduction = "umap.harmony", group.by = g_col)
  p1 + p2
} else {
  DimPlot(sc, reduction = "umap", group.by = g_col)
}
```


## Cell type annotations (maybe link MapMyCells)
Cell type identification is a class unto itself. Usually good methods involve a mix of automated cell type predictions, followed by careful curation to:

1. Verify cell type predictions & consolidate them across clusters
2. Eliminate heterotypic doublets & low quality cells
3. Subcluster and define population subtypes.

One useful tool for lightweight annotation prediction is the [MapMyCells](https://knowledge.brain-map.org/mapmycells/process) web server. In our case, we already have expertly curated cell types, defined by the XXXXXX column, as well as our experimental phenotype XXXXXX
```{r, annot_plot}
# Selecting your umap to plot
if (integ) {
  red <- "umap.harmony"
} else {
  red <- "umap"
}

# plots by functional annotation
p1 <- DimPlot(sc, reduction = red) + 
  ggtitle("clusters") +
  theme(plot.title = element_text(hjust = 0.5))
p2 <- DimPlot(sc, reduction = red, group.by = "letter.idents")

p1 + p2

```

## Differential expression 
```{r, sc_de}
# which column contains cell types?
ctype <- 'seurat_clusters'
# which column contains experimental groups?
group <- 'groups'

# standard DE analysis --------------------------------------------------------
Idents(sc) <- ctype
# running in a lapply loop to test each cell type (future.lapply can be even faster) 
de <- lapply(levels(sc), function(i) {
  data.table(
    FindMarkers(sc,
      ident.1 = 'g1',
      ident.2 = 'g2',
      group.by = group,
      subset.ident = i,
      test.use = "MAST",
      max.cells.per.ident = 10000
    )
  )
})
# Using data.table allows us to quickly/safely combine our de results into one table 
setattr(de, 'names', levels(sc))
de <- rbindlist(de, idcol = ctype)

```

While that can quickly get your result, too many cells from the same sample can result in inflated confidence in significance values. This can be partially limited with `max.cells.per.ident`, the preferred method is to pseudobulk your counts by sample for a bulkRNAseq style measure with DESeq2. However, this can be an issue if you have few samples per group. Try for n >= 6 per group.
```{r, sc_de_pseudo}
# using pseudobulk analysis to control for within-sample correlation -----------
pseudo <- AggregateExpression(sc, 
  assays = "RNA", 
  return.seurat = TRUE, 
  group.by = c("letter.idents", group, ctype)
)

Idents(pseudo) <- ctype
# running in a lapply loop to test each cell type (future.lapply can be even faster) 
de <- lapply(levels(pseudo), function(i) {
  data.table(
    FindMarkers(pseudo,
      ident.1 = 'g1',
      ident.2 = 'g2',
      group.by = group,
      subset.ident = i,
      min.cells.group = 2,
      test.use = "DESeq2"
    )
  )
})
# Using data.table allows us to quickly/safely combine our de results into one table 
setattr(de, 'names', levels(pseudo))
de <- rbindlist(de, idcol = ctype)
# ggplot facet volcanoes

```


# Introduction spatial (lecture) (from Sue)
Maybe just link to the broader [introduction](https://smcclatchy.github.io/spatial-transcriptomics/introduction.html) page?

# Spatial analysis (interactive)

## Download the data
## Clean and annotate
## Cell type annotations
## Spatial genes
## Spatial proximity
## Interaction changed features

# Single-cell x Spatial analysis (interactive)
1.	Differential ligand-receptor analysis
2.	Differential spatial relationships
3.	(maybe) Correlation to TREAT-AD

# Conclusion concepts (lecture)
(Maybe with time) Future promising tools (lecture or demo, non-interactive)
## Spacia + downstream genes in target cell
## Multimodal integration tools
